submissions/issue-16/milestone-1/Coelacanth/NFTStandard.cdc

```diff
@@ -0,0 +1,326 @@
+/*
+	This project is addressing problem "New Standard: NFT metadata #16"
+	The below code is untested and is meant as an illustration of the concept.
+	
+	
+	MetaDataUtil is added as a new default contract.
+	
+	Metadata is instantiated in the form of structs at the time the NFT is minted.
+	Metadata is organized by string tags, and each metadata element may have multiple tags allowing NFTs to conform to multiple tag schemas.
+	Metadata elements cannot be added after the NFT is minted, but Mutable elements can allow data to be modifiable.
+	
+	MetaDataElements are wrappers around two possible types of structs, one of which is immutable and the other of which is Mutable.
+	Immutable elements can use default types or be defined in custom structs that implement ITaggedMetaData and IImmutableMetaData. 
+		Defining a custom class allows the data elements to be immutable while allowing a developer to upgrade the contract that defines the struct to alter the tags.
+	Mutable elements store Capability pointers to IMetaDataProvider instances that can be stored anywhere.
+		Mutable elements can be used for metadata that can be modified, such as for leveling up a character.
+		Mutable elements can also be used for metadata that is shared between multiple NFTs, to minimize redundant data storage.
+			A Side-effect (benefit?) here is that if the instance is stored in a contract in the developers account, the developer will need to cover the metaData storage costs instead of the NFT holder.
+			PNG_RemoteDefaultImage is a default implementation that can be used to give an NFT a shared and/or updatable image.
+				RemotePNGProvider is an example of an implementation of IMetaDataProvider that allows multiple NFTs to share the same image via Capability reference.
+*/
+
+pub contract MetaDataUtil {
+    pub struct interface ITaggedMetaData {
+        pub fun getTags() : [String]
+    }
+
+    pub struct interface IMetaDataProvider {
+        pub fun getData() : AnyStruct
+        pub fun getDataType() : String
+    }
+
+    pub struct interface IImmutableMetaData {
+        pub let data: AnyStruct
+        pub let type: String
+    }
+
+    pub struct interface IMutableMetaData {
+        pub let dataProvider: Capability<&AnyStruct{IMetaDataProvider}>
+    }
+
+    pub struct MetaData {
+        pub let elements : [MetaDataElement]
```

Comment:
Note that the content of this array can be manipulated by the owner of the MetaData struct, and functions called on value types within it (and on types in arrays/dictionaries within those, and so on recursively...).

I usually recommend using `access(self) for the metadata stored in an object, and returning (an implicitly lazy copy of) it with a `pub` accessor.

i.e.:
```
access(self) let elements: [MetaDataElement]

pub fun getElements(): [MetaDataElement] {
  return self.elements
}
```

This applies to all the types below that expose a dictionary or array as a `pub` member variable.

Reply:
Thanks! I had gone through looking for that issue and changed a few others, but missed that one!